### PR TITLE
Refactor GitHub Integration

### LINKED ISSUE #6712: Refactor GitHub Integration
Some problems surfacing here

- Inconsistent contracts exposed due to `GitHubUser` & `GitHubApp` classes. Would be better to expose something like,
   ```js
   GitHub.authenticateAsUser(code, state)
   GitHub.authenticateAsInstallation(installationId)
   ```
   This goes well with the [documentation](https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/about-authentication-with-a-github-app).
- It'd be good to have transformers to transform GitHub API responses to data format we need. Currently, this is all being done ad-hoc within `unfurl`. This would yield itself unmaintainable if we decide to support unfurling for more resources(commits, comments etc.) going forward.

### PR DESCRIPTION
**Pull request by apoorv-mishra**
Closes #6712

### CODE REVIEW

File: plugins/github/server/uninstall.ts [hunk 1]
```diff
@@ -9,7 +9,8 @@ export async function uninstall(
     const installationId = integration.settings?.github?.installation.id;
 
     if (installationId) {
-      return githubApp.deleteInstallation(installationId);
+      const client = await GitHub.authenticateAsInstallation(installationId);
+      await GitHub.requestAppUninstall(installationId, client);
```
Comment (tommoor) [inline plugins/github/server/uninstall.ts:13]: This wrapper has ended up kind of unnecessary, and it's a bit of an awkward DX to pass `client` as second param – could just use octokit directly here?

File: plugins/github/server/github.ts [hunk 2]
```diff
@@ -1,73 +1,295 @@
...
-  /** Required to authenticate as GitHub App */
-  private static id = env.GITHUB_APP_ID;
-  private static key = env.GITHUB_APP_PRIVATE_KEY
+enum Resource {
+  PR = "pull",
+  Issue = "issue",
+}
+
+type PreviewData = {
+  [Resource.PR]: {
```
Comment (tommoor) [inline plugins/github/server/github.ts:21]: You can use the `UnfurlResponse` type for these, it's generic so the shape of `meta` can be passed in

File: plugins/github/server/api/github.ts [hunk 3]
```diff
@@ -65,13 +66,14 @@ router.get(
       }
     }
 
-    const githubUser = new GitHubUser({ code: code!, state: teamId });
+    const client = await GitHub.authenticateAsUser(code!, teamId);
+    const installationsByUser = await GitHub.requestAppInstallations(client);
```
Comment (tommoor) [inline plugins/github/server/api/github.ts:70]: Isn't this less performant than before? Why change to loading all installations 🤔

File: plugins/github/server/uninstall.ts (hunk 1 above)
Comment (apoorv-mishra) [reply plugins/github/server/uninstall.ts:13]: Can use directly, yes. I just stuck to keeping all `octokit.request()` calls within the class with the intent of not having to import Octokit in other places. Just trying to set a convention but flexible, yes.

File: plugins/github/server/api/github.ts (hunk 3 above)
Comment (apoorv-mishra) [reply plugins/github/server/api/github.ts:70]: We were always loading all the installations. There doesn't seem to be an API to load a single installation for a given user.

File: plugins/github/server/uninstall.ts (hunk 1 above)
Comment (tommoor) [reply plugins/github/server/uninstall.ts:13]: Lets use this pattern and extend the client perhaps:

https://github.com/octokit/core.js/#defaults

Will allow doing `await client.requestAppUninstall()`

**Review by tommoor — APPROVED** [review]

File: plugins/github/server/github.ts (hunk 2 above)
Comment (apoorv-mishra) [reply plugins/github/server/github.ts:21]: Resolving types separately altogether in another pass.

**Pull request merged**